net: IPv4 link local addresses support

locomuco opened the pull request:
this PR targets ZEP-803 (now https://github.com/zephyrproject-rtos/zephyr/issues/2310)

- [x] doc to be updated
- [x] conflict defending tbd

tbursztyka commented on include/net/ipv4_autoconf.h:
```diff
@@ -0,0 +1,47 @@
+
+/** Current state of IPv4 Autoconfiguration */
+enum net_ipv4_autoconf_state
+{
```
the { goes in front of the enum net_ipv4_autoconf_state

tbursztyka commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -34,6 +34,12 @@ config NET_DHCPV4
+config NET_IPV4_AUTOCONF
+    bool "Enable IPV4 autoconfiguration"
+    depends on NET_IPV4
+    depends on NET_ARP
```
you can coalesce the 2 depends on lines with &&

tbursztyka commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,288 @@
+/* delay between successive attempts */
+#define IPV4_AUTOCONF_RATE_LIMIT_INTERVAL 60
+/* minimum interval between defensive ARPs */
+#define IPV4_AUTOCONF_DEFEND_INTERVAL 10
```
all the above macros could go into subsys/net/ip/ipv4.h

tbursztyka commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,288 @@
+	struct net_arp_hdr *hdr;
+	struct net_eth_hdr *eth;
+
+	buf = net_nbuf_get_reserve_tx(0, K_FOREVER);
```
which version of Zephyr are you working on? We don't have nbuf api anymore, it's net_pkt now.

locomuco replied:
I did the implementation on a previous version, still need to pull up the APIs

tbursztyka commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,288 @@
+	default:
+		break;
+	}
+	return NET_DROP;
```
break a line after a } (unless it's another } etc...) apply this rule everywhere

tbursztyka commented on include/net/net_if.h:
```diff
@@ -271,6 +274,28 @@ struct net_if {
+        uint8_t announce_cnt;
+        /** Incoming conflict count */
+        uint8_t conflict_cnt;
+        /** Requested IP addr */
```
2 times the same comment on requested IP addr. 

Btw, couldn't current_ip be removed and you would use one of the address slot in ipv4 struct above?

tbursztyka commented on include/net/net_if.h:
```diff
@@ -271,6 +274,28 @@ struct net_if {
+        enum net_ipv4_autoconf_state state;
+    } ipv4_autoconf;
+    /** Timer for IPv4 autoconf requests (PROBE, ANNOUNCE) */
+    struct k_delayed_work ipv4_autoconf_timeout;
```
why is this out of ipv4_autoconf?

tbursztyka commented on include/net/net_if.h:
```diff
@@ -271,6 +274,28 @@ struct net_if {
+        *  address allocation.
+        */
+        enum net_ipv4_autoconf_state state;
+    } ipv4_autoconf;
```
I think you can move ipv4_autoconf into the above ipv4 struct. This is part of ipv4.

tbursztyka commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,288 @@
+	hdr->hwlen = sizeof(struct net_eth_addr);
+	hdr->protolen = sizeof(struct in_addr);
+	hdr->opcode = htons(NET_ARP_REQUEST);
+
```
all the above is not necessary, this is handled in susbsys/net/ip/l2/: arp.c and ethernet.c

locomuco replied:
can you add more details to that? how should the flow look like for that pkts?

jukkar replied:
I think this needs to be done here so that the code can prepare to send probe etc.

tbursztyka commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,288 @@
+		NET_DBG("Invalid ARP header (len %zu, min %zu bytes)",
+			net_buf_frags_len(buf),
+			sizeof(struct net_arp_hdr) - net_nbuf_ll_reserve(buf));
+		return NET_DROP;
```
same

jukkar commented on samples/net/ipv4_autoconf/src/main.c:
```diff
@@ -0,0 +1,78 @@
+{
+	NET_INFO("In main");
+
+	k_thread_create(&thread_data, &thread_stack[0], STACKSIZE,
```
Do we need a separate thread here, could we run sample directly from main()?
Some other samples are doing similar things (creating a thread that starts things) but we are trying to migrate away from doing it like this.

jukkar commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -34,6 +34,11 @@ config NET_DHCPV4
 	depends on NET_IPV4
 	default n
 
+config NET_IPV4_AUTOCONF
```
As you are introducing a new kconfig option for net, could you also add that option to tests/net/all/prj.conf file which we are using as a rudimentary way to confirm that there are no compile errors if that option is enabled.

locomuco replied:
done

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright (c) 2017 Matthias Boesl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
I wonder if it is enough if we have
SPDX-License-Identifier: Apache-2.0
line here instead of full license text. Just pondering this thought here, no need to change this if you think you want to have full text here.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+ */
+
+#if defined(CONFIG_NET_DEBUG_IPV4_AUTOCONF)
+#define SYS_LOG_DOMAIN "net/ipv4_autoconf"
```
I think we could have "net/ipv4ll" here instead, it is shorter which is good and still tells what functionality this line corresponds to.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+#include <net/net_if.h>
+
+#include <net/ipv4_autoconf.h>
+
```
Remove extra empty line.

locomuco replied:
done

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+}
+
+enum net_verdict net_ipv4_autoconf_input(struct net_if *iface,
+				struct net_pkt *pkt)
```
Indentation looks wrong.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+		if (iface->ipv4.conflict_cnt >=
+		    IPV4_AUTOCONF_MAX_CONFLICTS) {
+			NET_DBG("rate limiting\n");
+			return IPV4_AUTOCONF_RATE_LIMIT_INTERVAL *
```
If IPV4_AUTOCONF_RATE_LIMIT_INTERVAL is in seconds, then just do
K_SECONDS(IPV4_AUTOCONF_RATE_LIMIT_INTERVAL)
here so there would be no need to multiply by MSEC_PER_SEC.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+			return IPV4_AUTOCONF_ANNOUNCE_INTERVAL *
+			       MSEC_PER_SEC;
+		}
+		NET_DBG("delay 1s\n");
```
Do not put \n here as that would be added automatically by lower level logging macros. Now you are printing double \n

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+		       CONFIG_SYS_CLOCK_TICKS_PER_SEC +
+		       (sys_rand32_get() % MSEC_PER_SEC);
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		return IPV4_AUTOCONF_ANNOUNCE_INTERVAL *
```
K_SECONDS(...) here

jukkar commented on subsys/net/ip/l2/ethernet.c:
```diff
@@ -139,6 +142,9 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 			net_sprint_ll_addr((u8_t *)hdr->src.addr,
 					   sizeof(struct net_eth_addr)));
+#ifdef CONFIG_NET_IPV4_AUTOCONF
+		net_ipv4_autoconf_input(iface, pkt);
```
Is this intentional that we ignore the return value?
So regardless of what happens in net_ipv4_autoconf_input(), we still call net_arp_input()?

locomuco replied:
yes, this is basically just for conflict detection/resolution, the regular arp handling should still stay in place afterwards

jukkar commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -34,6 +34,11 @@ config NET_DHCPV4
+config NET_IPV4_AUTOCONF
+    bool "Enable IPV4 autoconfiguration"
+    depends on NET_IPV4 && NET_ARP
+    default n
```
You could mention the RFC 3927 here in the help text, and possible restrictions, or something that is not implemented from that rfc, etc.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+
+#if NET_DEBUG > 0
+	do {
+		NET_DBG("conflict detected from %s [%s] for %s", out,
```
Where is the "out" variable defined, I am not seeing it in this function?

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+	iface->ipv4.probe_cnt++;
+	iface->ipv4.state = NET_IPV4_AUTOCONF_PROBE;
+
+	k_delayed_work_init(&iface->ipv4.ipv4_autoconf_timeout,
```
Normally you should only call k_delayed_work_init() once if the callback does not change, I see now init called in send_announcement() too.

locomuco replied:
removed in announcements, should be fine, since there cannot be any announcements without probing beforehand

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,280 @@
+	}
+
+	NET_DBG("Announcing");
+	net_if_queue_tx(iface, announce);
```
I wonder if we could let the ethernet.c to actually send the packet here. So could we call net_send_data() here instead, which would properly do various checks etc before actually sending the pkt?

tbursztyka replied:
I fully agree here, and I don't like to see L2 specific code being cold in L3 (net_arp_prepare).
The other way round would be fine: ethernet could check if ipv4 autoconf is on going, and thus react accordingly with arp. (setting up the proper ipv4 address in arp request etc..)

jukkar reviewed (changes requested):
Lot of comments from me, I might have missed something thou so be prepared for further changes.

dbkinder commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,25 @@
+.. _ipv4-autoconf-sample:
```
Since this is a WIP, I'll assume this placeholder documentation will be expanded on before getting merged.

locomuco replied:
done

dbkinder reviewed (changes requested):
WIP Placeholder doc could use more information before getting merged :)

jukkar commented:
Ping @locomuco This has been pending long time, do you have time to work on the remaining issues. It would be great to be able to get this support to zephyr.

locomuco commented:
@jukkar I've been about to test it on real HW, currently I'm stuck getting the enc28j60 driver running with the new SPI interface on a stm32 platform, I might be able to resolve the blocking issues this weekend...

mbolivar commented:
> currently I'm stuck getting the enc28j60 driver running with the new SPI interface on a stm32 platform

@locomuco, in case it's useful, I also ran into some issues with the SPI driver which I've got patches up for in https://github.com/zephyrproject-rtos/zephyr/pull/832

locomuco commented:
@mbolivar thanks I've seen that, that was my plan for the weekend

locomuco commented:
@jukkar rebased

jukkar commented on include/net/ipv4_autoconf.h:
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright (c) 2017 Matthias Boesl
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
```
You could simplify the license boiler plate text here by just saying "SPDX-License-Identifier: Apache-2.0", like the other header files are doing.

jukkar commented on include/net/ipv4_autoconf.h:
```diff
@@ -0,0 +1,65 @@
+void net_ipv4_autoconf_start(struct net_if *iface);
+
+enum net_verdict net_ipv4_autoconf_input(struct net_if *iface,
+					 struct net_pkt *pkt);
```
Indent looks off-by-one in this line (params should be aligned), could also be the github viewer issue.

jukkar commented on include/net/net_if.h:
```diff
@@ -422,7 +458,7 @@ static inline enum net_verdict net_if_recv_data(struct net_if *iface,
  */
 static inline u16_t net_if_get_ll_reserve(struct net_if *iface,
-					  const struct in6_addr *dst_ip6)
+					     const struct in6_addr *dst_ip6)
```
Why do you change this line?

jukkar commented on include/net/net_if.h:
```diff
@@ -1077,7 +1113,7 @@ static inline void net_ipv6_set_hop_limit(struct net_if *iface,
  */
 static inline void net_if_ipv6_set_base_reachable_time(struct net_if *iface,
-						       u32_t reachable_time)
+						    u32_t reachable_time)
```
Same here

jukkar commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,62 @@
+Building and Running
+********************
+
+These are instructions for how to use this sample application running on :ref:frdm_k64f board to configure a link local IPv4 address and connect to a Linux host.
```
This looks too long line, could you split it. For generated documentation this does not matter but if reading via command line, this matters. Check also rest of the file for too long (>80 chars) lines.

jukkar commented on samples/net/ipv4_autoconf/prj_frdm_k64f.conf:
```diff
@@ -0,0 +1,24 @@
+
+CONFIG_NET_IPV4=y
+CONFIG_NET_IPV4_AUTO=y
+
```
Remove extra empty lines at the end of file.

jukkar commented on samples/net/ipv4_autoconf/prj_qemu_x86.conf:
```diff
@@ -0,0 +1,37 @@
+CONFIG_NETWORKING=y
```
It looks like you should be able to merge the prj files together and just have one prj.conf file. Note that you do not need to mention CONFIG_NET_SLIP_TAP in conf file as that will be automatically applied if compiled for qemu.

jukkar commented on samples/net/ipv4_autoconf/src/main.c:
```diff
@@ -0,0 +1,65 @@
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+		char buf[NET_IPV4_ADDR_LEN];
+
+		if (iface->config.ip.ipv4->unicast[i].addr_type != NET_ADDR_AUTOCONF) {
```
Please search and replace all "iface->config" by a call to "net_if_get_config(iface)" so instead of iface->config.ip.ipv4... we would have net_if_get_config(iface)->ip.ipv4...

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+
+	iface->config.ipv4auto.conflict_cnt++;
+
+	switch (iface->config.ipv4auto.state) {
```
Instead of iface->config, do net_if_get_config(iface). Do the change to rest of the file too.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+				    iface, &iface->config.ipv4auto.requested_ip)) {
+				NET_DBG("Failed to remove addr from iface");
+			}
+			/* restart probing after second conflict */
```
Empty line after }

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+			   IPV4_AUTOCONF_PROBE_NUM) {
+			return K_SECONDS(IPV4_AUTOCONF_ANNOUNCE_INTERVAL);
+		}
+		return IPV4_AUTOCONF_PROBE_WAIT *
```
Empty line after }

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+		return IPV4_AUTOCONF_PROBE_WAIT *
+		       CONFIG_SYS_CLOCK_TICKS_PER_SEC +
+		       (sys_rand32_get() % MSEC_PER_SEC);
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
```
Empty line before case xxx:

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+		       (sys_rand32_get() % MSEC_PER_SEC);
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		return K_SECONDS(IPV4_AUTOCONF_ANNOUNCE_INTERVAL);
+	default:
```
Empty line before default:

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+		(uint8_t)(sys_rand32_get() % 254);
+	iface->config.ipv4auto.requested_ip.s4_addr[3] =
+		(uint8_t)(sys_rand32_get() % 254);
+	send_probe(iface);
```
Would look better if there would be one empty line before send_probe();

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+	iface->config.ipv4auto.requested_ip.s4_addr[3] =
+		(uint8_t)(sys_rand32_get() % 254);
+	send_probe(iface);
+}
```
\n is missing at the end of the file.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,263 @@
+			send_probe(iface);
+			break;
+		}
+
```
Can you add comment here if this is intentional that we pass through without a break. So simple /* passthrough */ is enough, the idea is that anyone reading the code does not need to start to investigate if there is break missing or not.

jukkar commented on subsys/net/ip/l2/ethernet.c:
```diff
@@ -153,6 +156,9 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
 			net_sprint_ll_addr((u8_t *)hdr->src.addr,
 					   sizeof(struct net_eth_addr)));
+#ifdef CONFIG_NET_IPV4_AUTO
+		net_ipv4_autoconf_input(iface, pkt);
```
Shouln't we return the verdict to the caller here, so "return net_ipv4_autoconf_input(...)", now we always do arp thing regardless of autoconf return value?

locomuco replied:
not sure, I think the pkt should in any case reach the arp handler in the next function call after this one, can you add more details?

jukkar replied:
I am just asking this, the net_ipv4_autoconf_input() line just looked suspicious. If you are sure this is ok, then no worries here then.

locomuco replied:
should be fine as is

jukkar reviewed (changes requested):
Some comments, I might have missed some thou.

jukkar commented on include/net/net_if.h:
```diff
@@ -421,7 +457,7 @@ static inline enum net_verdict net_if_recv_data(struct net_if *iface,
  */
 static inline u16_t net_if_get_ll_reserve(struct net_if *iface,
-					  const struct in6_addr *dst_ip6)
+					 const struct in6_addr *dst_ip6)
```
As I mentioned already earlier, this change is not related to this LL support, please remove this change.

jukkar commented on include/net/net_if.h:
```diff
@@ -1076,7 +1112,7 @@ static inline void net_ipv6_set_hop_limit(struct net_if *iface,
  */
 static inline void net_if_ipv6_set_base_reachable_time(struct net_if *iface,
-						       u32_t reachable_time)
+								u32_t reachable_time)
```
Same here

jukkar commented:
@tbursztyka & @dbkinder please review this PR as you had change requests earlier.

jukkar reviewed (dismissed):
Sorry for long delay when reviewing this PR. LGTM.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,267 @@
+				    iface, &net_if_get_config(iface)->ipv4auto.requested_ip)) {
+				NET_DBG("Failed to remove addr from iface");
+			}
+			/* restart probing after second conflict */
```
Missed this one earlier, please put empty line after }

jukkar reviewed (changes requested):
Noticed some minor things when testing this so these could be good to fix too.

jukkar commented on samples/net/ipv4_autoconf/src/main.c:
```diff
@@ -0,0 +1,65 @@
+
+	iface = net_if_get_default();
+
+	net_ipv4_autoconf_start(iface);
```
The autoconf start could be placed in separate patch as that is not really related to actual autoconf implementation.

locomuco commented:
@jukkar changes applied and rebased

jukkar commented on subsys/net/ip/Kconfig.ipv4:
```diff
@@ -63,6 +71,13 @@ config NET_DEBUG_DHCPV4
 
+config NET_DEBUG_IPV4_AUTOCONF
+	bool "Debug IPv4 autoconf client"
+	depends on NET_IPV4_AUTOCONF
```
Noticed that I cannot enable debugging, the reason was that the depends on statement is wrong here, it should be "depends on NET_IPV4_AUTO"

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+	net_if_get_config(iface)->ipv4auto.probe_cnt = 0;
+	net_if_get_config(iface)->ipv4auto.announce_cnt = 0;
+	net_if_get_config(iface)->ipv4auto.conflict_cnt = 0;
+	memset(&net_if_get_config(iface)->ipv4auto.current_ip, 0, sizeof(struct in_addr));
```
My editor complains about this line, it is too long. Normally checkpatch/sanitycheck complains about these but for some reason it did not do that this time.

Perhaps you could do

`struct net_if_config *config = net_if_get_config(iface);
`

at the beginning of the func, and then use config var later in func. That way the lines get shorter and it looks a bit better.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+		(uint8_t)(sys_rand32_get() % 254);
+	net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[3] =
+		(uint8_t)(sys_rand32_get() % 254);
+
```
I suggest here a debug print like this

```
	NET_DBG("Starting probe for 169.254.%d.%d",
		net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[2],
		net_if_get_config(iface)->ipv4auto.requested_ip.s4_addr[3]);

```

otherwise it is a bit difficult to see what is being done.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+	       sizeof(struct net_eth_addr));
+	memset(&hdr->dst_hwaddr.addr, 0x00, sizeof(struct net_eth_addr));
+
+	net_ipaddr_copy(&hdr->dst_ipaddr, &net_if_get_config(iface)->ipv4auto.current_ip);
```
too long lines here, I suggest you use helper to get a config and then use the config var so that the lines will fit 80 chars

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+	arp_hdr = NET_ARP_HDR(pkt);
+
+	if (!net_ipv4_addr_cmp(&arp_hdr->dst_ipaddr,
+			       &net_if_get_config(iface)->ipv4auto.requested_ip)) {
```
too long line

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+		} else {
+			/* unset host ip */
+			if (!net_if_ipv4_addr_rm(
+				    iface, &net_if_get_config(iface)->ipv4auto.requested_ip)) {
```
too long line

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+	case NET_IPV4_AUTOCONF_ANNOUNCE:
+		if (net_if_get_config(iface)->ipv4auto.announce_cnt <=
+		    IPV4_AUTOCONF_ANNOUNCE_NUM - 1) {
+			net_ipaddr_copy(&net_if_get_config(iface)->ipv4auto.current_ip,
```
too long line

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+
+		/* success, add new IPv4 address */
+		if (!net_if_ipv4_addr_add(iface,
+					  &net_if_get_config(iface)->ipv4auto.requested_ip,
```
too long line

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,270 @@
+			return;
+		}
+
+		net_if_get_config(iface)->ipv4auto.state = NET_IPV4_AUTOCONF_CONFIGURED;
```
too long line

jukkar reviewed (commented):
I am testing this against a windows machine, and noticed that zephyr changes the LL IPv4 address very often, meaning that the net_ipv4_autoconf_start() is called many times which causes the IP address change.
This affects connectivity with peer device. Why do we change the IP address if there is no conflict?
I am looking the ethernet.c which calls net_ipv4_autoconf_input() for every received ARP packet. This in turn causes IP address change if we are in PROBE state. But it is quite normal thing for a peer to send ARP to zephyr in which case we should not change the IP address, or am I missing something here.

jukkar commented:
ping @locomuco Any chance you have time to look this?

locomuco commented:
@jukkar cannot promise, but I try to get it fixed on the weekend

locomuco commented:
@jukkar all style/config issues are fixed.
I couldn't reproduce your reportings regarding the conflict handling

this part of the code should ensure we don't react on regular arps
```C
	if (!net_ipv4_addr_cmp(&arp_hdr->dst_ipaddr,
		&net_if_get_config(iface)->ipv4auto.requested_ip)) {
		/* No conflict */
		return NET_DROP;
	}
```

besides that, I made the probing start depending on Interface UP management event, shich is not 100% correct, is there a way to get a notification about the link state of the ethernet interface, if not, I would set the PR to blocked until done so?

jukkar commented:
> I couldn't reproduce your reportings regarding the conflict handling

Thanks for the update. I will re-test this with windows, perhaps this was just something related to my setup.

jukkar commented:
Tentatively setting milestone to 1.13 release.

locomuco commented:
@jukkar after rebasing this to  #8027, link change propagation doesn't seem to be stable, any idea?

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -245,23 +253,44 @@ static void send_announcement(struct net_if *iface)
+		return;
+	}
+
 	/* initialize interface and start probing */
```
I think you need to also check NET_EVENT_IF_UP here as the event number is not a bitmask.

jukkar commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -245,23 +253,44 @@ static void send_announcement(struct net_if *iface)
+		    struct net_if *iface)
 {
+
+	if ((mgmt_event & NET_EVENT_IF_DOWN) &&
```
Check mgmt_event == NET_EVENT_IF_DOWN instead here.

jukkar reviewed (commented):
I see lot of white space issues, is you editor using different tab size than expected (should be 8). You could use uncrustify to make sources conform to style guide.

locomuco commented:
@jukkar style changes are applied, rebased to the latest dhcp renewal PR,
there is still one particular problem with the management API.

once an IP address is added using ```net_if_ipv4_addr_add``` the link up/down callbacks are not working any more.

The code works fine, when ```net_if_ipv4_addr_add``` is commented out.

does ```net_if_ipv4_addr_add``` need to be decoupled somehow, since it's generating a managemnt event by itself?

locomuco commented:
@jukkar after talking back to @tbursztyka the autoconf implementation now hooks directly into net_if

fully functional now, and ready for review

jukkar reviewed (changes requested):
Noticed some issues

pfalcon commented on subsys/net/ip/ipv4_autoconf.c:
```diff
@@ -0,0 +1,315 @@
+		}
+
+		return IPV4_AUTOCONF_PROBE_WAIT *
+		       CONFIG_SYS_CLOCK_TICKS_PER_SEC +
```
Why ticks are used at all, instead of milliseconds?

jukkar commented:
I tested this with #7251 and net-usb, works well with Windows. There are couple of comments that should be addressed and then we should be ready to merge this.

locomuco commented:
@jukkar can you double-check the timing change, otherwise the PR should be ready to merge

jukkar reviewed (approved):
LGTM, also tested this after latest changes and did not see any runtime issues.

jukkar commented:
@tbursztyka & @dbkinder could you revisit this PR so that we could get this merged?

dbkinder commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,66 @@
+
+To verify the Zephyr application is running and has configured an ip address
+type:
+$ ping -I eth1 169.254.218.128
```
add a code block for this command:

    To verify the Zephyr application is running and has configured an ip address
    type:

    .. code-block:: bash

       $ ping -I eth1 169.254.218.128

locomuco commented:
@dbkinder done

dbkinder reviewed (approved):
Thanks.

locomuco commented:
@tbursztyka can you do a final check on that PR?

locomuco commented:
@jukkar is there anything holding that PR back from merge

tbursztyka commented on samples/net/ipv4_autoconf/src/main.c:
```diff
@@ -0,0 +1,61 @@
+	for (i = 0; i < NET_IF_MAX_IPV4_ADDR; i++) {
+		char buf[NET_IPV4_ADDR_LEN];
+
+		if (net_if_get_config(iface)->ip.ipv4->unicast[i].addr_type != NET_ADDR_AUTOCONF) {
```
80 chars limit

tbursztyka reviewed (changes requested):
I stick to my comment about mixing up arp/eth stuff into ipv4: that must not happen.

locomuco commented:
@tbursztyka no problem, do you have a specific idea, how this should be restructured?

jukkar commented:
> @tbursztyka no problem, do you have a specific idea, how this should be restructured?

I cannot speak on behalf of @tbursztyka but one thing that could be separated is the ARP generation function prepare_arp(). At least it could be generalized and placed into arp.c.

tbursztyka commented:
@locomuco 100% of the logic you need from arp is already in arp code. There (in L2 in general) it is totally fine to have some checks on IP from a given net_pkt. So you could check "is length 0? is ipv4 autoconf (verify some state)? and act on it (do your arp probing). Afaik, pkt length is not tested in net_if_send_data so it's fine to have a net_pkt of AF_IPV4 family, without any content. Up to you to go on ethernet l2, do what's meant  to be done (mostly in arp.c). (if autoconf, it could pick up the address you want to give from net_pkt on its relevant struct net_if_ipv4_autoconf, and generate as usual the broadcast arp header. Done).

The current ethernet L2 is messy, but part of your feature has to go there and only there. Or then it's going to be even messier.

jukkar commented:
Please all note that ipv4 link-local address support is also applicable for 802.11 (wifi) and not just ethernet. 
The rfc specifically says that

> This specification applies to all IEEE 802 Local Area Networks (LANs) [802], including Ethernet [802.3], Token-Ring [802.5] and IEEE 802.11 wireless LANs [802.11], as well as to other link-layer technologies that operate at data rates of at least 1 Mbps, have a round-trip latency of at most one second, and support ARP [RFC826].

tbursztyka commented:
@jukkar sure but support arp only on ethernet atm, thus why it's located there. Once we get a new L2 requiring it, it will be put away from ethernet and inserted at l2 root dir.

jukkar commented:
Updated the code with new commit that addresses various issues that @tbursztyka had. The two commits will be merged together before merging the feature to upstream.

tbursztyka commented on include/net/net_if.h:
```diff
@@ -280,13 +280,13 @@ struct net_if_dhcpv4 {
 
 	/** Incoming conflict count */
-	uint8_t conflict_cnt;
+	u8_t conflict_cnt;
```
Actually, there is a hole in the original struct, could you move these 3 attributes to the end of the struct?

Also: the k_delayed_work, could you apply the same as I did for dhcpv4 which you adapted as well in lldp? Better now than later.

tbursztyka reviewed (changes requested):
There is still some work to do to get it right.

jukkar committed:
net: IPv4 link local support

Add basic IPv4 Link Local support as described in RFC 3927.

Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>
Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
samples: net: IPv4 autoconf sample application

Adding IPv4 autoconf sample application that can be used to
test IPv4 autoconf functionality.

Signed-off-by: Matthias Boesl <matthias.boesl@gmail.com>
Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar commented:
New version that fixes the sanity errors

tbursztyka commented on subsys/net/l2/ethernet/arp.c:
```diff
@@ -304,15 +311,30 @@ static inline struct net_pkt *arp_prepare(struct net_if *iface,
 	} else {
-		my_addr = &NET_IPV4_HDR(pending)->src;
+		if (is_ipv4_auto && IS_ENABLED(CONFIG_NET_IPV4_AUTO)) {
+#if defined(CONFIG_NET_IPV4_AUTO)
```
This is awful: IS_ENABLED _and_ #ifdef used side by side? 
To avoid that, make inline function returning pointer on ipv4auto address in net_if, if not enabled: just make stub ones that return NULL.

tbursztyka reviewed (changes requested):
I am still not fully happy, there should be a way to let arp knows that ipv4 auto is running, without passing an argument to the function (which makes it annoying when ipv4 auto is not built in). You could have a net_if function telling you true/false that it is actually probing/running or something like that.
(if so, actually it could even be possible to let ethernet/arp do its job without calling net_arp_prepare from ipv4 auto code directly).

Anyway, one comment in the code.

jukkar commented:
> there should be a way to let arp knows that ipv4 auto is running, without passing an argument to the function (which makes it annoying when ipv4 auto is not built in). You could have a net_if function telling you true/false that it is actually probing/running or something like that.

It is not very obvious to me how this is possible. The IPv4 autoconf is running in different level than ARP or net_if so those lower entities do not know anything about IPv4 autoconf.

tbursztyka reviewed (approved):
ok sold, let's not spend more time on this.

pfalcon commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,69 @@
+********
+
+This application starts a IPv4 autoconf and self-assigns
+a random IPv4 address in the 169.254.0.0/16 range, it defends
```
@jukkar, @dbkinder : FYI, I think the documentation for this sample needs cleanup. E.g. I doubt "defends" is the right word here.

locomuco replied:
@pfalcon, wording matches with RFC descriptions, should be ok as is 
see https://tools.ietf.org/html/rfc3927#section-2

pfalcon replied:
@locomuco : Thanks for response, and indeed, interesting wording is in the RFC ;-). I'd still think that outside that RFC, it sounds a bit strange. There're few other minor issues with the docs, which I didn't mark explicitly but meant when talked about need for cleanup.

pfalcon commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,69 @@
+This application starts a IPv4 autoconf and self-assigns
+a random IPv4 address in the 169.254.0.0/16 range, it defends
+the IPv4 address and resolves IPv4 conflicts if multiple
+parties try to allocate an identical address
```
@locomuco : Re: previous comment: full stop missing at the end of sentence.

pfalcon commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,69 @@
+connect to a Linux host.
+
+Connect ethernet cable from :ref:Freedom-K64F board <frdm_k64f> to Linux
+host machine and check for new interfaces:
```
Strange ":" at the end of sentence.

pfalcon commented on samples/net/ipv4_autoconf/README.rst:
```diff
@@ -0,0 +1,69 @@
+Connect ethernet cable from :ref:Freedom-K64F board <frdm_k64f> to Linux
+host machine and check for new interfaces:
+
+Running Avahi client in Linux Host
```
"*on* Linux host" would sound better.
